### ISSUE TITLE
The SPA default page middleware could not return the default page '/index.html' in production application

### ISSUE THREAD

**Issue by danobri** [aspnet/JavaScriptServices#1729]
I have an ASP.Net Core / Angular 5 application that is logging errors to the Windows Event Log in our production environment.  The following error is getting logged frequently and intermittently in production, but I have no idea why.  The application is definitely in production mode, has been published, and is functioning.

`System.InvalidOperationException: The SPA default page middleware could not return the default page '/index.html' because it was not found, and no other middleware handled the request.
Your application is running in Production mode, so make sure it has been published, or that you have built your SPA manually. Alternatively you may wish to switch to the Development environment.`

AFAIK no users are experiencing a problem with the site when these errors occur.  Any idea why these errors are being thrown, and whether there is in fact a problem that needs to be addressed?

**Comment by Spaier** [comment]
Do angular output directory and specified path in `services.AddSpaStaticFiles(configuration => configuration.RootPath = $"your/output/path");` match?

**Comment by danobri** [comment]
I am not explicitly setting the RootPath:

`app.UseSpaStaticFiles();`

But I am setting the SourcePath:

` app.UseSpa(spa =>
      {
        spa.Options.SourcePath = "ClientApp";
...
`

Do I need to also set the RootPath like in your example?

And here is what's in my .csproj file:

```xml
<PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <TypeScriptCompileBlocked>true</TypeScriptCompileBlocked>
    <IsPackable>false</IsPackable>
    <SpaRoot>ClientApp\</SpaRoot>
    <DefaultItemExcludes>$(DefaultItemExcludes);$(SpaRoot)node_modules\**</DefaultItemExcludes>
    <!-- Set this to true if you enable server-side prerendering -->
    <BuildServerSideRenderer>false</BuildServerSideRenderer>
</PropertyGroup>

<Target Name="PublishRunWebpack" AfterTargets="ComputeFilesToPublish">
    <!-- As part of publishing, ensure the JS resources are freshly built in production mode -->
    <Exec WorkingDirectory="$(SpaRoot)" Command="npm install" />
    <Exec WorkingDirectory="$(SpaRoot)" Command="npm run build -- --prod" />
    <Exec WorkingDirectory="$(SpaRoot)" Command="npm run build:ssr -- --prod" Condition=" '$(BuildServerSideRenderer)' == 'true' " />

    <!-- Include the newly-built files in the publish output -->
    <ItemGroup>
      <DistFiles Include="$(SpaRoot)dist\**; $(SpaRoot)dist-server\**" />
      <DistFiles Include="$(SpaRoot)node_modules\**" Condition="'$(BuildServerSideRenderer)' == 'true'" />
      <ResolvedFileToPublish Include="@(DistFiles->'%(FullPath)')" Exclude="@(ResolvedFileToPublish)">
        <RelativePath>%(DistFiles.Identity)</RelativePath>
        <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
      </ResolvedFileToPublish>
    </ItemGroup>
  </Target>
```

I'm wondering if my problem is that I still have the SSR build instructions in the csproj even though SSR is not enabled, and dist-server does not exist..?

**Comment by Spaier** [comment]
SSR isn't the problem. You should set path in `ConfigureServices()` method via `services.AddSpaStaticFiles(configuration => configuration.RootPath = $"ClientApp/dist");`. Then in `Configure` you shouldn't specify source path:
```cs
app.UseSpaStaticFiles();
app.UseSpa(configuration => { /*spa.Options.SourcePath = "ClientApp";*/ });
```

**Comment by danobri** [comment]
Sorry - not sure if you saw my edit, but I was in fact setting the RootPath correctly, I was just looking in the wrong place. :)  I will try removing the SourcePath setting, however, I'm fairly certain this configuration came with the dotnet new angular template.  What is the SourcePath setting used for and why shouldn't it be set?

**Comment by Spaier** [comment]
This is used to determine folder where package.json scripts should be executed.
Like `new AngularCliBuilder(npmScript: "build:ssr")` or ` spa.UseAngularCliServer(npmScript: "start");`.
```cs
app.UseSpa(spa =>
{
    spa.Options.SourcePath = "ClientApp";

    spa.UseSpaPrerendering(options =>
    {
        options.BootModulePath = $"{spa.Options.SourcePath}/dist-server/main.bundle.js";
        options.BootModuleBuilder = env.IsDevelopment()
            ? new AngularCliBuilder(npmScript: "build:ssr")
            : null;
        options.ExcludeUrls = new[] { "/sockjs-node" };
    });

    if (env.IsDevelopment())
    {
        spa.UseAngularCliServer(npmScript: "start");
    }
});
```

**Comment by danobri** [comment]
Thanks - I am using a proxy:

```c#
spa.UseProxyToSpaDevelopmentServer("https://localhost:4200");
```
So sounds like I probably don't need the SourcePath setting.  That said - are you saying the SourcePath setting is the cause of the errors in my production environment?

**Comment by Spaier** [comment]
No, `SourcePath` does nothing if you use proxy during development. Application probably can't find index.html because it wasn't specified like this: `services.AddSpaStaticFiles(configuration => configuration.RootPath = $"ClientApp/dist");`

**Comment by danobri** [comment]
I think maybe you aren't seeing the edits I made to my original response.  I am setting the RootPath as suggested.  (Maybe try refreshing the this Github page to ensure you are seeing the latest content).

**Comment by Spaier** [comment]
Now I see it. Can you show what's inside of `UseSpa()`?

**Comment by danobri** [comment]
Not much...  Will of course remove the SourcePath setting since it's not needed, but this is what is currently in production:

```c#
app.UseSpa(spa =>
{
  spa.Options.SourcePath = "ClientApp";
  
  if (env.IsDevelopment())
  {
    spa.UseProxyToSpaDevelopmentServer("https://localhost:4200");
  }
});
```

**Comment by Spaier** [comment]
Is index.html is present after build in ClientApp/dist? I'd try to remove node_modules and install again, remove `SourcePath` and redeploy.

**Comment by danobri** [comment]
Index.html is present in clientapp/dist, otherwise the application wouldn't function at all - that is what is confusing about the error message.  The error is being logged in a production application that seems to working fine.  I will try removing the node_modules folder locally, and then rebuild and republish and see if that makes a difference.  What makes you think that will resolve the error?

**Comment by danobri** [comment]
Removing the node_modules folder and republishing did not solve the issue - still seeing quite a few of these errors logged on the production server.  Any idea why that would be?

**Comment by joshberry** [comment]
I'm seeing the same issue as @danobri on a production site that appears to be working fine otherwise. The exceptions are intermittent, but persistent. I'm using the Angular spa template pretty much straight away with SSR enabled.

**Comment by joshpearce** [comment]
It's an IIS Default Document issue. Just add the following to your web.config:

```
<system.webServer>
   ...
      <defaultDocument enabled="false" />
</system.webServer>
```

**Comment by waynebrantley** [comment]
So, it was our vulnerability scanner that caused the exceptions to be generated.  This means I can now reproduce it by scanning the IP this core app is installed on.   I tried the ``defaultDocument enabled=false`` in web.config, with no success.  (running scanner gave errors again)

Any ideas @SteveSandersonMS ?

**Comment by grascioni** [comment]
I'm also getting this error randomly in a couple of production sites that work just fine.

**Comment by maxisam** [comment]
Here is our case for past 2 days. It is really random. We have 3 servers and one of them barely has this issue in last 48 hours but it did have a lot before that. 

![image](https://user-images.githubusercontent.com/456807/46151942-82e68200-c235-11e8-9e80-9be61115eff8.png)

**Comment by vertonghenb** [comment]
Using React here and publishing the dotnetcore app 2.1 as a Windows Service
I believe it's because CRA creates an additional folder structure 
ClientApp/build/...
Therefore the index.html file isn't found.

**Comment by mjtpena** [comment]
This happened to my Azure App Service. All I had to do was to restart the service. Worked like charm.

**Comment by addikhann** [comment]
check angular.json file use  "outputPath": "dist",

And Startup file

 services.AddSpaStaticFiles(configuration =>
      {
        configuration.RootPath = "ClientApp/dist";
      });

**Comment by CodySchrank** [comment]
I can't even get a fresh `dotnet new react` to start in production without it throwing this exception:
`System.InvalidOperationException: The SPA default page middleware could not return the default page '/index.html' because it was not found, and no other middleware handled the request...`

Seems like there is definitely a bug in the building process of the spa somewhere?

**Comment by Eilon** [comment]
@ryanbrandenburg / @mkArtakMSFT - any ideas?

**Comment by JohnGalt1717** [comment]
So the reason we're getting this is someone is trying to POST AND OPTIONS to /index.html not GET. Which causes this error message and causes a 500 error.

Of course of Azure/IIS sees too many errors in too short a time it tries and recycles the app. the problem is that it's doing it over and over again which causes everything in that App Service Plan to crash. :<

How do we properly handle this without the site throwing a 500 error?

**Comment by joshberry** [comment]
Would be nice if this wasn't the default behavior. I was able to prevent the exceptions using a rewrite rule in my web.config.

```
<rewrite>
  <rules>
    <rule name="Prevent requests to index">
      <match url="^index\.html$" />
      <action type="Redirect" url="/" redirectType="Permanent" appendQueryString="true" />
    </rule>
  </rules>
</rewrite>
```

**Comment by robertpmansion** [comment]
I also had this issue what I did is use the rewrite middleware by doing the below.

```
app.UseRewriter(new RewriteOptions()
                .AddRedirect("index.html", "/"));
```

**Comment by rachelfreeman** [comment]
I would really like to see the root of this problem solved. We have a .net core 2.1 app with the latest React version, no web config since we are using json files for environment variables. We are randomly getting this error several times at once. Most of the time it's fine and then randomly it will happen a dozen times. The redirect seems to just be a work around and not solve the root issue.

**Comment by yakeer** [comment]
Anyone has got a solution for this?
I'm facing the same issue, can't get the app deployed successfully..

**Comment by waynebrantley** [comment]
Seems like these javascript services are unsupported?
@SteveSandersonMS can you advise what course of action to take?

**Comment by jraadt** [comment]
I had 1000's of these errors within the first hour after we deployed to Azure App Service. By removing this line from MyProject.csproj the errors stopped and everything kept working.

`<PackageReference Include="Microsoft.AspNetCore.SpaServices.Extensions" Version="2.1.1" />`

**Comment by joshberry** [comment]
@jraadt If you removed the SpaServices extensions, how are you doing the SPA setup in Startup.cs? Using the default template, your app should no longer build after removing that package reference.

**Comment by jraadt** [comment]
@joshberry I agree it makes no sense. As a sanity test I created a new project and did the following steps:

1. `dotnet new angular`
2. `dotnet restore`
3. `dotnet run`
4. I opened a web browser to verify the default ASP.NET Core/Angular app is running fine.
5. Delete `<PackageReference Include="Microsoft.AspNetCore.SpaServices.Extensions" Version="2.1.1" />` from the csproj file.
6. `dotnet restore`
7. `dotnet run`
8. Everything still built fine and I refreshed my web browser and verified the app still runs just fine. Incremented counter and everything.

I wonder if it's included in the Microsoft.AspNetCore.App package. I'd love some clarification from the Microsoft team on this.

**Comment by Eilon** [comment]
`Microsoft.AspNetCore.SpaServices.Extensions` was indeed in `Microsoft.AspNetCore.App` for 2.1, so that's why the explicit reference wasn't needed:
* https://github.com/aspnet/Universe/blob/2.1.5/build/artifacts.props#L165

But I don't think it was in there in the 2.0.x world because it came out after 2.0.0.

**Comment by waynebrantley** [comment]
How does any of that explain the errors stopping?  Any idea?

**Comment by waynebrantley** [comment]
I feel this error is not about finding ever user that sends OPTIONS to your endpoint and having them stop.  That this issue has not been addressed is crazy.   We get it every week due to vulnerability scanners.

**Comment by MogauGeeky** [comment]
Hi guys, I just started using Azure App Service and my deployment is throwing the same error. Unlike others who can still access the app mine crashes completely. I noticed from the console that visual studio deploy place the Spa files 3 folders deep, I think this might be the issue, I'm not sure how to make the files only be put within the dist folder.
![image](https://user-images.githubusercontent.com/10864511/49888297-47784e80-fe47-11e8-8e4b-2ad37bb5a780.png)

Is there anyone else who has the same folder structure withing their deployment?
